pkg/querier/queryrange/codec.go

```diff
@@ -823,6 +823,103 @@ func mapToPrometheusResponse(mergedResponse *logproto.VolumeResponse) queryrange
 	}
 }
 
+func mergeOrderedStreams(resps []*LokiResponse, limit uint32, direction logproto.Direction) []logproto.Stream {
```

Comment:
The existing function `mergeOrderedNonOverlappingStreams` is not used anywhere any more (except in tests), right? Why not replace it?

Reply:
Have you looked at the merge implementation using the heap sort? I would prefer fixing the heap sort rather than writing another merge algorithm (assuming there is a bug in the heap sort).

Reply:
We could also think about doing the sorting using a loser tree (see `pkg/util/loser/tree.go`)

Reply:
> We could also think about doing the sorting using a loser tree (see `pkg/util/loser/tree.go`)

I haven't thought about loser tree, Thanks for your suggestion！I will have a look at it and to improve the implementation If it have better performance.

Reply:
Finally I implement it again using loser tree, Thanks! 
Here is the benchmark compare with old method of limited query:
goos: darwin
goarch: arm64
pkg: github.com/grafana/loki/pkg/querier/queryrange
BenchmarkResponseMerge
BenchmarkResponseMerge/mergeOrderedStreams_limited
BenchmarkResponseMerge/mergeOrderedStreams_limited-8  	                               2000	  18819018 ns/op
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_limited
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_limited-8        2000	  15761584 ns/op

Reply:
@chaudum Can this be continue? Let me know if there is any problem.

Reply:
@honganan Sorry for not coming back to this earlier. I just ran the benchmarks, and they don't look good tbh.

**old**

```console
$ gotest -v -count=5 -test.run=^$ -benchtime=100x -test.bench=BenchmarkResponseMerge/mergeOrdered ./pkg/querier/queryrange/... 
goos: linux
goarch: amd64
pkg: github.com/grafana/loki/pkg/querier/queryrange
cpu: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
BenchmarkResponseMerge
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_unlimited
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_unlimited-8         	     100	    976852 ns/op
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_unlimited-8         	     100	    940463 ns/op
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_unlimited-8         	     100	    947441 ns/op
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_unlimited-8         	     100	    951037 ns/op
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_unlimited-8         	     100	    926886 ns/op
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_limited
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_limited-8           	     100	  17564476 ns/op
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_limited-8           	     100	  17543280 ns/op
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_limited-8           	     100	  17722991 ns/op
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_limited-8           	     100	  18168282 ns/op
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_limited-8           	     100	  17690048 ns/op
```

**new**
```console
$ gotest -v -count=5 -test.run=^$ -benchtime=100x -test.bench=BenchmarkResponseMerge/mergeOrderedStreams ./pkg/querier/queryrange/...
goos: linux
goarch: amd64
pkg: github.com/grafana/loki/pkg/querier/queryrange
cpu: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
BenchmarkResponseMerge
BenchmarkResponseMerge/mergeOrderedStreams_unlimited
BenchmarkResponseMerge/mergeOrderedStreams_unlimited-8         	     100	  20367019 ns/op
BenchmarkResponseMerge/mergeOrderedStreams_unlimited-8         	     100	  20324013 ns/op
BenchmarkResponseMerge/mergeOrderedStreams_unlimited-8         	     100	  20894591 ns/op
BenchmarkResponseMerge/mergeOrderedStreams_unlimited-8         	     100	  20863402 ns/op
BenchmarkResponseMerge/mergeOrderedStreams_unlimited-8         	     100	  20715545 ns/op
BenchmarkResponseMerge/mergeOrderedStreams_limited
BenchmarkResponseMerge/mergeOrderedStreams_limited-8           	     100	  20312703 ns/op
BenchmarkResponseMerge/mergeOrderedStreams_limited-8           	     100	  20645201 ns/op
BenchmarkResponseMerge/mergeOrderedStreams_limited-8           	     100	  20557591 ns/op
BenchmarkResponseMerge/mergeOrderedStreams_limited-8           	     100	  20753145 ns/op
BenchmarkResponseMerge/mergeOrderedStreams_limited-8           	     100	  20346076 ns/op
```

Had to change the benchmark name for `benchstat`:

```console
$ benchstat old.txt new.txt 
old.txt:5: missing iteration count
old.txt:6: missing iteration count
old.txt:12: missing iteration count
new.txt:5: missing iteration count
new.txt:6: missing iteration count
new.txt:12: missing iteration count
goos: linux
goarch: amd64
pkg: github.com/grafana/loki/pkg/querier/queryrange
cpu: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
                                              │   old.txt    │                 new.txt                 │
                                              │    sec/op    │     sec/op      vs base                 │
ResponseMerge/mergeOrderedStreams_unlimited-8   961.1µ ± ∞ ¹   19838.8µ ± ∞ ¹  +1964.15% (p=0.008 n=5)
ResponseMerge/mergeOrderedStreams_limited-8     17.40m ± ∞ ¹     20.24m ± ∞ ¹    +16.35% (p=0.008 n=5)
geomean                                         4.089m           20.04m         +390.06%
¹ need >= 6 samples for confidence interval at level 0.95
```

Reply:
It may lack a shortcut in `unlimited` case, I will add it recent. 
In `limited` case the increasing elapsed seems reasonable as the old implementation does not do the right things.

Reply:
After adding a shortcut, cost have small increasing than old, benchmarks from me:
```
goos: darwin
goarch: arm64
pkg: github.com/grafana/loki/pkg/querier/queryrange
                                              │ /Users/honganan/bench/0.txt │      /Users/honganan/bench/3.txt      │
                                              │           sec/op            │    sec/op     vs base                 │
ResponseMerge/mergeStreams_unlimited-8                         38.88m ± ∞ ¹   38.15m ± ∞ ¹        ~ (p=1.000 n=1) ²
ResponseMerge/mergeOrderedStreams_unlimited-8                  1.202m ± ∞ ¹   1.998m ± ∞ ¹        ~ (p=1.000 n=1) ²
ResponseMerge/mergeStreams_limited-8                           38.62m ± ∞ ¹   38.29m ± ∞ ¹        ~ (p=1.000 n=1) ²
ResponseMerge/mergeOrderedStreams_limited-8                    16.33m ± ∞ ¹   25.44m ± ∞ ¹        ~ (p=1.000 n=1) ²
geomean                                                        13.10m         16.51m        +25.99%
¹ need >= 6 samples for confidence interval at level 0.95
² need >= 4 samples to detect a difference at alpha level 0.05
```

And memory allocation seems much less than before:
![image](https://github.com/grafana/loki/assets/4892550/05be645e-8ca7-44d8-9adb-869ba88da44d)

Reply:
@chaudum What's your opinion? I think skip log lines is a serious problem, and if the fix is not good enough I am glad to see a better one.

Reply:
@chaudum Please have a look when you have time

Reply:
@chaudum Sorry for the incorrect comparison of the benchmark. I ran it again, and there was some increase in time consumption(Partly because the old method not doing the right things). Here is the result:
**old**
```
$ go test -v -count=5 -test.run=^$ -benchtime=100x -test.bench=BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams ./pkg/querier/queryrange/...
goos: darwin
goarch: arm64
pkg: github.com/grafana/loki/pkg/querier/queryrange
BenchmarkResponseMerge
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_unlimited
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_unlimited-8                 100           1524674 ns/op
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_unlimited-8                 100           1408833 ns/op
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_unlimited-8                 100           1695640 ns/op
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_unlimited-8                 100           1448439 ns/op
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_unlimited-8                 100           1476367 ns/op
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_limited
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_limited-8                   100          16939148 ns/op
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_limited-8                   100          16831235 ns/op
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_limited-8                   100          16605767 ns/op
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_limited-8                   100          16837140 ns/op
BenchmarkResponseMerge/mergeOrderedNonOverlappingStreams_limited-8                   100          16961015 ns/op
PASS
ok      github.com/grafana/loki/pkg/querier/queryrange  10.163s
?       github.com/grafana/loki/pkg/querier/queryrange/limits   [no test files]
PASS
ok      github.com/grafana/loki/pkg/querier/queryrange/queryrangebase   0.610s
?       github.com/grafana/loki/pkg/querier/queryrange/queryrangebase/definitions       [no test files]
```
**new**
```
$ go test -v -count=5 -test.run=^$ -benchtime=100x -test.bench=BenchmarkResponseMerge/mergeOrderedStreams ./pkg/querier/queryrange/...              
goos: darwin
goarch: arm64
pkg: github.com/grafana/loki/pkg/querier/queryrange
BenchmarkResponseMerge
BenchmarkResponseMerge/mergeOrderedStreams_unlimited
BenchmarkResponseMerge/mergeOrderedStreams_unlimited-8               100           2899758 ns/op
BenchmarkResponseMerge/mergeOrderedStreams_unlimited-8               100           2859770 ns/op
BenchmarkResponseMerge/mergeOrderedStreams_unlimited-8               100           2821386 ns/op
BenchmarkResponseMerge/mergeOrderedStreams_unlimited-8               100           2864822 ns/op
BenchmarkResponseMerge/mergeOrderedStreams_unlimited-8               100           2823145 ns/op
BenchmarkResponseMerge/mergeOrderedStreams_limited
BenchmarkResponseMerge/mergeOrderedStreams_limited-8                 100          27421963 ns/op
BenchmarkResponseMerge/mergeOrderedStreams_limited-8                 100          27226710 ns/op
BenchmarkResponseMerge/mergeOrderedStreams_limited-8                 100          27246525 ns/op
BenchmarkResponseMerge/mergeOrderedStreams_limited-8                 100          27418054 ns/op
BenchmarkResponseMerge/mergeOrderedStreams_limited-8                 100          28732325 ns/op
PASS
ok      github.com/grafana/loki/pkg/querier/queryrange  16.823s
PASS
ok      github.com/grafana/loki/pkg/querier/queryrange/queryrangebase   0.688s
?       github.com/grafana/loki/pkg/querier/queryrange/queryrangebase/definitions       [no test files]
```
**compare**
```
benchstat benchmark-old.txt benchmark-new2.txt
benchmark-old.txt:5: missing iteration count
benchmark-old.txt:6: missing iteration count
benchmark-old.txt:12: missing iteration count
benchmark-new2.txt:5: missing iteration count
benchmark-new2.txt:6: missing iteration count
benchmark-new2.txt:12: missing iteration count
goos: darwin
goarch: arm64
pkg: github.com/grafana/loki/pkg/querier/queryrange
                                              │ benchmark-old.txt │         benchmark-new2.txt          │
                                              │      sec/op       │    sec/op     vs base               │
ResponseMerge/mergeOrderedStreams_unlimited-8        1.476m ± ∞ ¹   2.860m ± ∞ ¹  +93.70% (p=0.008 n=5)
ResponseMerge/mergeOrderedStreams_limited-8          16.84m ± ∞ ¹   27.42m ± ∞ ¹  +62.84% (p=0.008 n=5)
geomean                                              4.986m         8.855m        +77.60%
¹ need >= 6 samples for confidence interval at level 0.95
```